core/src/main/java/org/pentaho/platform/scheduler2/action/SchedulerFilenameUtils.java

```diff
@@ -39,61 +38,126 @@ private SchedulerFilenameUtils() {
+   * Windows path separator, defined by {@link #WINDOWS_PATH_SEPARATOR}, is not supported.
    * <p/>
    *
    * Examples:
    * <p/>
    * <pre>
-   * /foo + bar                  -->   /foo/bar
-   * /foo/ + bar                 -->   /foo/bar
-   * /foo + /bar                 -->   /foo/bar
-   * /foo/ + /bar                -->   /foo/bar
-   * /foo + bar.txt              -->   /foo/bar.txt
-   * /foo/ + bar.txt             -->   /foo/bar.txt
-   * /foo + /bar.txt             -->   /foo/bar.txt
-   * /foo/ + /bar.txt            -->   /foo/bar.txt
-   * scheme://foo + bar          -->   scheme://foo/bar
-   * scheme://foo/ + bar         -->   scheme://foo/bar
-   * scheme://foo + /bar         -->   scheme://foo/bar
-   * scheme://foo/ + /bar        -->   scheme://foo/bar
-   * scheme://foo + bar.txt      -->   scheme://foo/bar.txt
-   * scheme://foo/ + bar.txt     -->   scheme://foo/bar.txt
-   * scheme://foo + /bar.txt     -->   scheme://foo/bar.txt
-   * scheme://foo/ + /bar.txt    -->   scheme://foo/bar.txt
+   * /foo + bar                       -->   /foo/bar
+   * /foo/ + bar                      -->   /foo/bar
+   * /foo + /bar                      -->   /foo/bar
+   * /foo/ + /bar                     -->   /foo/bar
+   * /foo + bar.txt                   -->   /foo/bar.txt
+   * /foo/ + bar.txt                  -->   /foo/bar.txt
+   * /foo + /bar.txt                  -->   /foo/bar.txt
+   * /foo/ + /bar.txt                 -->   /foo/bar.txt
+   * </pre>
+   * Same permutations for the following:
+   * <pre>
+   * scheme://foo/ + /bar.txt         -->   scheme://foo/bar.txt
+   * /foo/ + some/folder/path/bar.txt --> /foo/some/folder/path/bar.txt
    * </pre>
    *
    * @param basePath
    * @param pathToAdd
-   * @return combined path consisting of: <code>basePath</code> + {@value #PATH_SEPARATOR} + <code>pathToAdd</code>
+   * @return if both arguments are non-null then
+   * <p/> combined path consisting of: <code>basePath</code> + {@value #PATH_SEPARATOR} + <code>pathToAdd</code> <p/> , otherwise null.
+   * @throws IllegalArgumentException if path contains windows separator defined by {@link #WINDOWS_PATH_SEPARATOR}
    */
   public static String concat( String basePath, String pathToAdd ) {
 
     if ( basePath == null || pathToAdd == null ) {
       return null;
     }
+
+    checkPaths( basePath, pathToAdd );
+
     /*
      * NOTE don't alter schema syntax ie remove // in a URI.
      * Don't call any function that eventually calls org.apache.commons.io.FilenameUtils#normalize
      * Similar logic to org.pentaho.platform.web.http.api.resources.SchedulerOutputPathResolver
      */
-    return getNoEndSeparator( basePath )  + PATH_SEPARATOR + FilenameUtils.getName( pathToAdd );
+    return getNoEndSeparator( toSb( basePath ) )
+        .append( PATH_SEPARATOR )
+        .append( getNoEndSeparator( toSb( pathToAdd ).reverse() ).reverse() ) // remove separator at the start of path
```

Comment:
Decided not to create a new one line function. It still seems pretty readable. Depending on implementation of StringBuilder, `reverse()` can be O(1). Even if  `reverse()` is O(n), overall the concatenation of strings in O(n) for various reasons.